api/views/submission.py

```diff
@@ -778,3 +778,99 @@ def bulkcheck(self, request: Request) -> Response:
                 urls.pop(urls.index(submission.url))
 
         return Response(status=status.HTTP_200_OK, data=urls)
+
+    @csrf_exempt
+    @swagger_auto_schema(
+        manual_parameters=[
+            Parameter(
+                "user_id",
+                "query",
+                type="number",
+                description="The user to center the leaderboard on.",
+            ),
+            Parameter(
+                "top_count",
+                "query",
+                type="number",
+                description="The number of users to show from the top leaderboard.",
+            ),
+            Parameter(
+                "above_count",
+                "query",
+                type="number",
+                description="The number of users to show above the given user.",
+            ),
+            Parameter(
+                "below_count",
+                "query",
+                type="number",
+                description="The number of users to show below the given user.",
+            ),
+        ],
+        responses={404: "No volunteer with the specified ID."},
+    )
+    @action(detail=False, methods=["get"])
+    def leaderboard(self, request: Request,) -> Response:
+        """Get the leaderboard for the given user."""
+        user_id = request.GET.get("user_id", None)
+        if user_id is not None:
+            user_id = int(user_id)
+        top_count = int(request.GET.get("top_count", 5))
+        above_count = int(request.GET.get("above_count", 5))
+        below_count = int(request.GET.get("below_count", 5))
+
+        above_data = None
+        user_data = None
+        below_data = None
+
+        rank_query = (
+            # Apply the provided submission filters
+            self.filter_queryset(Submission.objects)
+            .filter(completed_by__isnull=False)
+            # Add author information
+            .select_related("completed_by")
+            # Group by author
+            .values(
+                "completed_by", "completed_by__username", "completed_by__date_joined"
+            )
+            # Count gamma
+            .annotate(
+                gamma=Count("completed_by"),
+                id=F("completed_by"),
+                username=F("completed_by__username"),
+                date_joined=F("completed_by__date_joined"),
+            )
+            .values("id", "username", "gamma", "date_joined")
+            .order_by(F("gamma").desc(), F("date_joined").desc())
+        )
+        # We can't filter the query further, because we'd lose the rank annotations
+        # We convert the query into a list to extract the necessary entries
+        # TODO: This is very inefficient, maybe there's a better way to do this?
+        query_list = list(rank_query)
```

Comment:
Expanding into a tuple should be slightly faster. `query_list = (*rank_query,)`

Reply:
I merged this with the rank annotations, so we can't use this approach unfortunately.